From 51b110b1a79b3ea3cac7f9e2972e4e131ce5235b Mon Sep 17 00:00:00 2001
From: Brendan King <Brendan.King@imgtec.com>
Date: Tue, 18 Jul 2023 18:13:27 +0100
Subject: [PATCH 02/71] dri2: add a driver compatibility check extension

Add a DRI extension, DRI_DRIVER_COMPATIBILITY, to determine the
compatibilty of the render and display GPUs. The extension consists
of a single function, which is only called if the render and display
GPU FDs are different. If the function is not present, and the FDs
are different, the GPUs are assumed to be incompatible, as before
(i.e. intermediate linear buffers and blits may be needed to
transfer data between them).

Upstream-Status: Inappropriate [oe specific]

---
 include/GL/internal/dri_interface.h     | 19 ++++++++++++
 src/egl/drivers/dri2/egl_dri2.c         | 30 +++++++++++++++----
 src/egl/drivers/dri2/egl_dri2.h         |  4 +++
 src/egl/drivers/dri2/platform_wayland.c | 19 ++++++------
 src/egl/drivers/dri2/platform_x11.c     |  4 +--
 src/glx/dri3_glx.c                      | 39 ++++++++++++++++++-------
 src/glx/dri3_priv.h                     |  4 +++
 7 files changed, 91 insertions(+), 28 deletions(-)

diff --git a/include/GL/internal/dri_interface.h b/include/GL/internal/dri_interface.h
index 7778241485c..0f2c0b2877c 100644
--- a/include/GL/internal/dri_interface.h
+++ b/include/GL/internal/dri_interface.h
@@ -2249,4 +2249,23 @@ struct __DRImutableRenderBufferLoaderExtensionRec {
                                void *loaderPrivate);
 };
 
+#define __DRI_DRIVER_COMPATIBILITY "DRI_DriverCompatibility"
+#define __DRI_DRIVER_COMPATIBILITY_VERSION 1
+
+typedef struct __DRIdriverCompatibilityExtensionRec __DRIdriverCompatibilityExtension;
+struct __DRIdriverCompatibilityExtensionRec {
+   __DRIextension base;
+
+   /**
+    * Check compatibility of render and display GPU drivers.
+    * Return true if the drivers can interoperate without restriction
+    * (e.g. no intermediate linear buffers or blits between buffers
+    * are required).
+    */
+   bool (*checkDriverCompatibility)(int fd_render_gpu,
+                                    const char *driver_name_render_gpu,
+                                    int fd_display_gpu,
+                                    const char *driver_name_display_gpu);
+};
+
 #endif
diff --git a/src/egl/drivers/dri2/egl_dri2.c b/src/egl/drivers/dri2/egl_dri2.c
index 88d5681ddfd..fb980f6127d 100644
--- a/src/egl/drivers/dri2/egl_dri2.c
+++ b/src/egl/drivers/dri2/egl_dri2.c
@@ -706,6 +706,8 @@ static const struct dri_extension_match dri3_driver_extensions[] = {
     false},
    {__DRI_CONFIG_OPTIONS, 2, offsetof(struct dri2_egl_display, configOptions),
     true},
+   { __DRI_DRIVER_COMPATIBILITY, 1, offsetof(struct dri2_egl_display,
+    driver_compatibility), true },
 };
 
 static const struct dri_extension_match dri2_driver_extensions[] = {
@@ -714,6 +716,8 @@ static const struct dri_extension_match dri2_driver_extensions[] = {
    {__DRI_DRI2, 4, offsetof(struct dri2_egl_display, dri2), false},
    {__DRI_CONFIG_OPTIONS, 2, offsetof(struct dri2_egl_display, configOptions),
     true},
+   { __DRI_DRIVER_COMPATIBILITY, 1, offsetof(struct dri2_egl_display,
+    driver_compatibility), true },
 };
 
 static const struct dri_extension_match dri2_core_extensions[] = {
@@ -1070,9 +1074,24 @@ dri2_create_screen(_EGLDisplay *disp)
    struct dri2_egl_display *dri2_dpy = dri2_egl_display(disp);
    char *driver_name_display_gpu;
 
-   if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu) {
+   if (dri2_dpy->fd_render_gpu == dri2_dpy->fd_display_gpu) {
+      dri2_dpy->compat_gpus = true;
+      driver_name_display_gpu = NULL;
+   } else {
       driver_name_display_gpu =
          loader_get_driver_for_fd(dri2_dpy->fd_display_gpu);
+
+      if (dri2_dpy->driver_compatibility)
+         dri2_dpy->compat_gpus =
+            dri2_dpy->driver_compatibility->checkDriverCompatibility(dri2_dpy->fd_render_gpu,
+                                                                     dri2_dpy->driver_name,
+                                                                     dri2_dpy->fd_display_gpu,
+                                                                     driver_name_display_gpu);
+      else
+         dri2_dpy->compat_gpus = false;
+   }
+
+   if (!dri2_dpy->compat_gpus) {
       if (driver_name_display_gpu) {
          /* check if driver name is matching so that non mesa drivers
           * will not crash.
@@ -1082,10 +1101,12 @@ dri2_create_screen(_EGLDisplay *disp)
                0, dri2_dpy->fd_display_gpu, dri2_dpy->loader_extensions,
                dri2_dpy->driver_extensions, &dri2_dpy->driver_configs, disp);
          }
-         free(driver_name_display_gpu);
       }
    }
 
+   if (driver_name_display_gpu)
+      free(driver_name_display_gpu);
+
    int screen_fd = dri2_dpy->swrast ? -1 : dri2_dpy->fd_render_gpu;
    dri2_dpy->dri_screen_render_gpu = dri2_dpy->mesa->createNewScreen(
       0, screen_fd, dri2_dpy->loader_extensions, dri2_dpy->driver_extensions,
@@ -1096,7 +1117,7 @@ dri2_create_screen(_EGLDisplay *disp)
       return EGL_FALSE;
    }
 
-   if (dri2_dpy->fd_render_gpu == dri2_dpy->fd_display_gpu)
+   if (dri2_dpy->compat_gpus)
       dri2_dpy->dri_screen_display_gpu = dri2_dpy->dri_screen_render_gpu;
 
    dri2_dpy->own_dri_screen = true;
@@ -1283,8 +1304,7 @@ dri2_display_destroy(_EGLDisplay *disp)
 
       dri2_dpy->core->destroyScreen(dri2_dpy->dri_screen_render_gpu);
 
-      if (dri2_dpy->dri_screen_display_gpu &&
-          dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu)
+      if (dri2_dpy->dri_screen_display_gpu && !dri2_dpy->compat_gpus)
          dri2_dpy->core->destroyScreen(dri2_dpy->dri_screen_display_gpu);
    }
    if (dri2_dpy->fd_display_gpu >= 0 &&
diff --git a/src/egl/drivers/dri2/egl_dri2.h b/src/egl/drivers/dri2/egl_dri2.h
index 4c29a50ff1a..46f8cdc6284 100644
--- a/src/egl/drivers/dri2/egl_dri2.h
+++ b/src/egl/drivers/dri2/egl_dri2.h
@@ -254,6 +254,7 @@ struct dri2_egl_display {
    const __DRI2interopExtension *interop;
    const __DRIconfigOptionsExtension *configOptions;
    const __DRImutableRenderBufferDriverExtension *mutable_render_buffer;
+   const __DRIdriverCompatibilityExtension *driver_compatibility;
    /* fd of the GPU used for rendering. */
    int fd_render_gpu;
    /* fd of the GPU used for display. If the same GPU is used for display
@@ -319,6 +320,9 @@ struct dri2_egl_display {
    /* gralloc vendor usage bit for front rendering */
    uint32_t front_rendering_usage;
 #endif
+
+   /* Indicates the render and display GPUs are compatible */
+   bool compat_gpus;
 };
 
 struct dri2_egl_context {
diff --git a/src/egl/drivers/dri2/platform_wayland.c b/src/egl/drivers/dri2/platform_wayland.c
index e0fe7b3ac9d..7a6c63669fc 100644
--- a/src/egl/drivers/dri2/platform_wayland.c
+++ b/src/egl/drivers/dri2/platform_wayland.c
@@ -997,7 +997,7 @@ create_dri_image_from_dmabuf_feedback(struct dri2_egl_surface *dri2_surf,
       dri2_surf->back->dri_image = loader_dri_create_image(
          dri2_dpy->dri_screen_render_gpu, dri2_dpy->image,
          dri2_surf->base.Width, dri2_surf->base.Height, dri_image_format,
-         (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu) ? 0 : flags,
+         (!dri2_dpy->compat_gpus) ? 0 : flags,
          modifiers, num_modifiers, NULL);
 
       if (dri2_surf->back->dri_image)
@@ -1033,7 +1033,7 @@ create_dri_image(struct dri2_egl_surface *dri2_surf,
    dri2_surf->back->dri_image = loader_dri_create_image(
       dri2_dpy->dri_screen_render_gpu, dri2_dpy->image, dri2_surf->base.Width,
       dri2_surf->base.Height, dri_image_format,
-      (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu) ? 0 : use_flags,
+      (!dri2_dpy->compat_gpus) ? 0 : use_flags,
       modifiers, num_modifiers, NULL);
 }
 
@@ -1108,7 +1108,7 @@ get_back_bo(struct dri2_egl_surface *dri2_surf)
       use_flags |= __DRI_IMAGE_USE_PROTECTED;
    }
 
-   if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu &&
+   if (!dri2_dpy->compat_gpus &&
        dri2_surf->back->linear_copy == NULL) {
       uint64_t linear_mod = DRM_FORMAT_MOD_LINEAR;
       __DRIimage *linear_copy_display_gpu_image = NULL;
@@ -1278,7 +1278,7 @@ update_buffers(struct dri2_egl_surface *dri2_surf)
           dri2_surf->color_buffers[i].age > BUFFER_TRIM_AGE_HYSTERESIS) {
          wl_buffer_destroy(dri2_surf->color_buffers[i].wl_buffer);
          dri2_dpy->image->destroyImage(dri2_surf->color_buffers[i].dri_image);
-         if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu)
+         if (!dri2_dpy->compat_gpus)
             dri2_dpy->image->destroyImage(
                dri2_surf->color_buffers[i].linear_copy);
          dri2_surf->color_buffers[i].wl_buffer = NULL;
@@ -1584,7 +1584,7 @@ dri2_wl_swap_buffers_with_damage(_EGLDisplay *disp, _EGLSurface *draw,
    if (!dri2_surf->current->wl_buffer) {
       __DRIimage *image;
 
-      if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu)
+      if (!dri2_dpy->compat_gpus)
          image = dri2_surf->current->linear_copy;
       else
          image = dri2_surf->current->dri_image;
@@ -1618,7 +1618,7 @@ dri2_wl_swap_buffers_with_damage(_EGLDisplay *disp, _EGLSurface *draw,
       wl_surface_damage(dri2_surf->wl_surface_wrapper, 0, 0, INT32_MAX,
                         INT32_MAX);
 
-   if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu) {
+   if (!dri2_dpy->compat_gpus) {
       _EGLContext *ctx = _eglGetCurrentContext();
       struct dri2_egl_context *dri2_ctx = dri2_egl_context(ctx);
       dri2_dpy->image->blitImage(
@@ -2061,7 +2061,7 @@ dri2_wl_add_configs_for_visuals(_EGLDisplay *disp)
          }
       }
 
-      if (!assigned && dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu) {
+      if (!assigned && !dri2_dpy->compat_gpus) {
          struct dri2_egl_config *dri2_conf;
          int alt_dri_image_format, c, s;
 
@@ -2257,8 +2257,7 @@ dri2_initialize_wayland_drm(_EGLDisplay *disp)
       }
    }
 
-   if (dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu &&
-       dri2_dpy->image->blitImage == NULL) {
+   if (!dri2_dpy->compat_gpus && dri2_dpy->image->blitImage == NULL) {
       _eglLog(_EGL_WARNING, "wayland-egl: Different GPU selected, but the "
                             "Image extension in the driver is not "
                             "compatible. blitImage() is required");
@@ -2275,7 +2274,7 @@ dri2_initialize_wayland_drm(_EGLDisplay *disp)
     * because the buffer of the EGLImage has likely a tiling mode the server
     * gpu won't support. These is no way to check for now. Thus do not support
     * the extension */
-   if (dri2_dpy->fd_render_gpu == dri2_dpy->fd_display_gpu)
+   if (dri2_dpy->compat_gpus)
       disp->Extensions.WL_create_wayland_buffer_from_image = EGL_TRUE;
 
    disp->Extensions.EXT_buffer_age = EGL_TRUE;
diff --git a/src/egl/drivers/dri2/platform_x11.c b/src/egl/drivers/dri2/platform_x11.c
index 6af987ed7ad..1b77adea07a 100644
--- a/src/egl/drivers/dri2/platform_x11.c
+++ b/src/egl/drivers/dri2/platform_x11.c
@@ -1545,7 +1545,7 @@ dri2_initialize_x11_swrast(_EGLDisplay *disp)
 #endif
       dri2_dpy->swap_available = EGL_TRUE;
       dri2_x11_setup_swap_interval(disp);
-      if (dri2_dpy->fd_render_gpu == dri2_dpy->fd_display_gpu)
+      if (dri2_dpy->compat_gpus)
          disp->Extensions.KHR_image_pixmap = EGL_TRUE;
       disp->Extensions.NOK_texture_from_pixmap = EGL_TRUE;
       disp->Extensions.CHROMIUM_sync_control = EGL_TRUE;
@@ -1625,7 +1625,7 @@ dri2_initialize_x11_dri3(_EGLDisplay *disp)
 
    dri2_x11_setup_swap_interval(disp);
 
-   if (dri2_dpy->fd_render_gpu == dri2_dpy->fd_display_gpu)
+   if (dri2_dpy->compat_gpus)
       disp->Extensions.KHR_image_pixmap = EGL_TRUE;
    disp->Extensions.NOK_texture_from_pixmap = EGL_TRUE;
    disp->Extensions.CHROMIUM_sync_control = EGL_TRUE;
diff --git a/src/glx/dri3_glx.c b/src/glx/dri3_glx.c
index a1c9fb56696..f479bd0bea3 100644
--- a/src/glx/dri3_glx.c
+++ b/src/glx/dri3_glx.c
@@ -589,7 +589,7 @@ dri3_destroy_screen(struct glx_screen *base)
    struct dri3_screen *psc = (struct dri3_screen *) base;
 
    /* Free the direct rendering per screen data */
-   if (psc->fd_render_gpu != psc->fd_display_gpu && psc->driScreenDisplayGPU) {
+   if (!psc->compat_gpus && psc->driScreenDisplayGPU) {
       loader_dri3_close_screen(psc->driScreenDisplayGPU);
       psc->core->destroyScreen(psc->driScreenDisplayGPU);
    }
@@ -738,7 +738,7 @@ dri3_bind_extensions(struct dri3_screen *psc, struct glx_display * priv,
        * can have a tiling mode we can't read. Thus we can't create
        * a texture from them.
        */
-      if (psc->fd_render_gpu == psc->fd_display_gpu &&
+      if (psc->compat_gpus &&
          (strcmp(extensions[i]->name, __DRI_TEX_BUFFER) == 0)) {
          psc->texBuffer = (__DRItexBufferExtension *) extensions[i];
          __glXEnableDirectExtension(&psc->base, "GLX_EXT_texture_from_pixmap");
@@ -843,12 +843,28 @@ dri3_create_screen(int screen, struct glx_display * priv)
        { __DRI_CORE, 1, offsetof(struct dri3_screen, core), false },
        { __DRI_IMAGE_DRIVER, 1, offsetof(struct dri3_screen, image_driver), false },
        { __DRI_MESA, 1, offsetof(struct dri3_screen, mesa), false },
+       { __DRI_DRIVER_COMPATIBILITY, 1, offsetof(struct dri3_screen, driver_compatibility), true },
    };
    if (!loader_bind_extensions(psc, exts, ARRAY_SIZE(exts), extensions))
       goto handle_error;
 
-   if (psc->fd_render_gpu != psc->fd_display_gpu) {
+   if (psc->fd_render_gpu == psc->fd_display_gpu) {
+      psc->compat_gpus = true;
+      driverNameDisplayGPU = NULL;
+   } else {
       driverNameDisplayGPU = loader_get_driver_for_fd(psc->fd_display_gpu);
+
+      if (psc->driver_compatibility)
+         psc->compat_gpus =
+            psc->driver_compatibility->checkDriverCompatibility(psc->fd_render_gpu,
+                                                                driverName,
+                                                                psc->fd_display_gpu,
+                                                                driverNameDisplayGPU);
+      else
+         psc->compat_gpus = false;
+   }
+
+   if (!psc->compat_gpus) {
       if (driverNameDisplayGPU) {
 
          /* check if driver name is matching so that non mesa drivers
@@ -863,11 +879,12 @@ dri3_create_screen(int screen, struct glx_display * priv)
                                                    extensions,
                                                    &driver_configs, psc);
          }
-
-         free(driverNameDisplayGPU);
       }
    }
 
+   if (driverNameDisplayGPU)
+      free(driverNameDisplayGPU);
+
    psc->driScreenRenderGPU =
       psc->image_driver->createNewScreen2(screen, psc->fd_render_gpu,
                                           pdp->loader_extensions,
@@ -879,7 +896,7 @@ dri3_create_screen(int screen, struct glx_display * priv)
       goto handle_error;
    }
 
-   if (psc->fd_render_gpu == psc->fd_display_gpu)
+   if (psc->compat_gpus)
       psc->driScreenDisplayGPU = psc->driScreenRenderGPU;
 
    dri3_bind_extensions(psc, priv, driverName);
@@ -894,17 +911,17 @@ dri3_create_screen(int screen, struct glx_display * priv)
       goto handle_error;
    }
 
-   if (psc->fd_render_gpu != psc->fd_display_gpu && psc->image->base.version < 9) {
+   if (!psc->compat_gpus && psc->image->base.version < 9) {
       ErrorMessageF("Different GPU, but image extension version 9 or later not found\n");
       goto handle_error;
    }
 
-   if (psc->fd_render_gpu != psc->fd_display_gpu && !psc->image->blitImage) {
+   if (!psc->compat_gpus && !psc->image->blitImage) {
       ErrorMessageF("Different GPU, but blitImage not implemented for this driver\n");
       goto handle_error;
    }
 
-   if (psc->fd_render_gpu == psc->fd_display_gpu && (
+   if (psc->compat_gpus && (
        !psc->texBuffer || psc->texBuffer->base.version < 2 ||
        !psc->texBuffer->setTexBuffer2
        )) {
@@ -998,7 +1015,7 @@ dri3_create_screen(int screen, struct glx_display * priv)
    InfoMessageF("Using DRI3 for screen %d\n", screen);
 
    psc->prefer_back_buffer_reuse = 1;
-   if (psc->fd_render_gpu != psc->fd_display_gpu && psc->rendererQuery) {
+   if (!psc->compat_gpus && psc->rendererQuery) {
       unsigned value;
       if (psc->rendererQuery->queryInteger(psc->driScreenRenderGPU,
                                            __DRI2_RENDERER_PREFER_BACK_BUFFER_REUSE,
@@ -1018,7 +1035,7 @@ handle_error:
    if (psc->driScreenRenderGPU)
        psc->core->destroyScreen(psc->driScreenRenderGPU);
    psc->driScreenRenderGPU = NULL;
-   if (psc->fd_render_gpu != psc->fd_display_gpu && psc->driScreenDisplayGPU)
+   if (!psc->compat_gpus && psc->driScreenDisplayGPU)
        psc->core->destroyScreen(psc->driScreenDisplayGPU);
    psc->driScreenDisplayGPU = NULL;
    if (psc->fd_display_gpu >= 0 && psc->fd_render_gpu != psc->fd_display_gpu)
diff --git a/src/glx/dri3_priv.h b/src/glx/dri3_priv.h
index 0a9c0a648f8..6b615ceecaf 100644
--- a/src/glx/dri3_priv.h
+++ b/src/glx/dri3_priv.h
@@ -97,6 +97,7 @@ struct dri3_screen {
    const __DRI2rendererQueryExtension *rendererQuery;
    const __DRI2interopExtension *interop;
    const __DRIconfig **driver_configs;
+   const __DRIdriverCompatibilityExtension *driver_compatibility;
 
    void *driver;
    /* fd of the GPU used for rendering. */
@@ -109,6 +110,9 @@ struct dri3_screen {
    bool prefer_back_buffer_reuse;
 
    struct loader_dri3_extensions loader_dri3_ext;
+
+   /* Indicates the render and display GPUs are compatible */
+   bool compat_gpus;
 };
 
 struct dri3_drawable {
-- 
2.34.1

